pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/actions/RestartDeclarativePipelineAction.java

```diff
@@ -103,7 +105,12 @@ public boolean isRestartEnabled() {
                 !run.isBuilding() &&
                 run.hasPermission(Item.BUILD) &&
                 run.getParent().isBuildable() &&
-                getExecution() != null;
+                getExecution() != null &&
+                isRestartableFromStage();
+    }
+
+    private boolean isRestartableFromStage() {
+        return run.getParent().getAction(DisableRestartFromStageAction.class) == null;
```

Comment:
nit: I would inline the method and delete it. `isRestartableFromStage` seems too generic of a name in this context and it seems like it would be confused with `isRestartEnabled`.

```suggestion
                run.getParent().getAction(DisableRestartFromStageAction.class) == null;
```

Reply:
Sounds good

Reply:
Please verify the change